### PR TITLE
Chrome: Adding the Permalink popup

### PR DESCRIPTION
**Pull request by youknowriad**
refs #577 

- This PRs adds a read-only permalink popup (for now).
- I'm also adding a generic `ClipboardButton` based on the `clipboard` package

### CODE REVIEW

**Comment by youknowriad** [comment]
Editing permalinks is not always authorised, a permalink format containing the "slug" is required to allow editing the slug.

AFAIK It's not possible to fetch the permalink format using the Rest API right now. Thoughts on this @nylen

**Comment by jasmussen** [comment]
NICE! Thanks for working on this.

Let's make a few visual improvements. I can help with this, and it can be in a separate PR.

For example, let's pretend the title field is a block. So it gets the same 2px gray block boundary as a selected block is. 

That lets us also position the permalink popup in the exact same coordinates as the block toolbars are. That is, indended from the left edge of the selected border the block padding distance, and 2px pushed down from the top border so it's "sitting on the border":

screen shot 2017-06-06 at 14 28 54

Can we change the text of the copy button to say "Copied!" when you click?

Can we make the permalink clickable, in addition to being editable?

**Comment by youknowriad** [comment]
@jasmussen Thanks for the review, I've tried to address the feedback as I could.

**Comment by jasmussen** [comment]
Love this! Really impressive work. Teensy nitpick — can we add the same fade animation to the title border as the normal blocks have? Otherwise, visually SOLID. Thank! 👍

**Comment by iseulde** [comment]
When I focus the title, then focus a block, and then again the title, some UI lingers.

screenshot 2017-06-07 13 54 02

File: components/clipboard-button/index.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	render() {
+		const { className, children } = this.props;
+		const classes = classnames( 'components-clipboard-button', className );
+
+		return (
+			<Button
+				ref={ ref => this.button = ref }
+				className={ classes }
```
Comment (iseulde) [inline components/clipboard-button/index.js]: It's a bit confusing that that there's no click handler here.

File: editor/post-title/index.js [hunk 2]
```diff
@@ -15,35 +18,95 @@ import { ENTER } from 'utils/keycodes';
...
+		this.textareaContainer = ref;
+	}
+
+	onSelectionChange() {
+		const textarea = this.textareaContainer.textarea;
+		if (
+			document.activeElement === textarea &&
+			textarea.selectionStart !== textarea.selectionEnd
+		) {
+			this.select();
```
Comment (iseulde) [inline editor/post-title/index.js]: Might it be better to name this `onSelect`?

File: editor/post-title/index.js [hunk 3]
```diff
@@ -15,35 +18,95 @@ import { ENTER } from 'utils/keycodes';
...
+		this.setState( { isSelected: true } );
+	}
 
-	const onKeyDown = ( event ) => {
+	unselect() {
+		this.setState( { isSelected: false } );
+	}
+
+	handleClickOutside() {
+		this.setState( { isSelected: false } );
```
Comment (iseulde) [inline editor/post-title/index.js]: Wouldn't `blur` work here?

File: components/clipboard-button/index.js (hunk 1 above)
Comment (youknowriad) [inline components/clipboard-button/index.js]: yep, that's how the `clipboard` library works.

File: editor/post-title/index.js (hunk 3 above)
Comment (youknowriad) [inline editor/post-title/index.js]: It doesn't because of the popup

**Review by ellatrix — APPROVED** [review]
Works nicely.

**Comment by aduth** [comment]
Created core ticket ( + patch ) for REST API `permalink_structure` setting:

https://core.trac.wordpress.org/ticket/41014

**Comment by jasmussen** [comment]
Did this regress? I don't see it in master anymore...

**Comment by youknowriad** [comment]
@jasmussen Did you save your post?

**Comment by jasmussen** [comment]
Ah of course, that's it.

File: editor/post-title/index.js [hunk 4]
```diff
@@ -3,47 +3,111 @@
...
+	componentWillUnmount() {
+		document.removeEventListener( 'selectionchange', this.onSelectionChange );
+	}
+
+	bindTextarea( ref ) {
+		this.textareaContainer = ref;
+	}
+
+	onSelectionChange() {
+		const textarea = this.textareaContainer.textarea;
```
Comment (aduth) [inline editor/post-title/index.js]: What is this doing? Can we add an inline comment?

Comment (aduth) [reply editor/post-title/index.js]: See #5422

**Pull request merged**